### ISSUE TITLE
get_url Fails to Validate Otherwise Valid HTTPS URL

### ISSUE THREAD

**Issue by naftulikay** [ansible/ansible#23642]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

`get_url` Module.

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.6.9 (unknown, Nov  9 2016, 21:08:27) [GCC 4.8.4]
```

Other Python versions, since I'm testing on Travis:

```
python version = 2.6.9 (unknown, Nov  9 2016, 21:08:27) [GCC 4.8.4]
python version = 2.7.12 (default, Aug 25 2016, 13:54:13) [GCC 4.8.4]
python version = 3.4.5 (default, Nov  9 2016, 21:08:56) [GCC 4.8.4]
python version = 3.5.2 (default, Aug 19 2016, 19:25:35) [GCC 4.8.4]
python version = 3.6.0 (default, Dec 23 2016, 14:52:51) [GCC 4.8.4]
```

##### CONFIGURATION

None.

##### OS / ENVIRONMENT

Travis CI on their default Precise (Ubuntu 12.04) and their beta Trusty (14.04) images.

##### SUMMARY

I'm [attempting to write an Ansible role to install Rust in a Vagrant VM](https://github.com/naftulikay/ansible-role-vagrant-rust-dev/pull/1):

```
- name: download rustup
  get_url:
    url: "https://sh.rustup.rs"
    dest: "{{ rustup_file }}"
    owner: "{{ vagrant_user }}"
    group: "{{ vagrant_group }}"
    mode: "0700"
  become: "{{ vagrant_user }}"
```

Running `curl` from within Travis:

```
$ curl -sSfv https://sh.rustup.rs
* About to connect() to sh.rustup.rs port 443 (#0)
*   Trying 54.215.0.2... connected
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, CERT (11):
* SSLv3, TLS handshake, Server key exchange (12):
* SSLv3, TLS handshake, Server finished (14):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using ECDHE-RSA-AES128-SHA
* Server certificate:
* 	 subject: CN=sh.rustup.rs
* 	 start date: 2017-04-03 01:57:00 GMT
* 	 expire date: 2017-07-02 01:57:00 GMT
* 	 subjectAltName: sh.rustup.rs matched
* 	 issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
* 	 SSL certificate verify ok.
> GET / HTTP/1.1
> User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
> Host: sh.rustup.rs
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx/1.4.6 (Ubuntu)
< Date: Sun, 16 Apr 2017 04:57:17 GMT
< Content-Type: text/x-sh
< Content-Length: 9319
< Connection: keep-alive
< Vary: Accept-Encoding
< x-amz-id-2: rCDFrdNGSviwXpP8LH5cJx3aByim7SeKa8wzBxSzk6yMBAe3pcUocb3NhI5Re08erOrK9WPYCU8=
< x-amz-request-id: E1DE57D3199C50FF
< Last-Modified: Sat, 08 Apr 2017 22:29:18 GMT
< ETag: "255da4e3a56b60b1adf3a159fc035912"
< x-amz-version-id: MTJMXVyeDGGgkJNPtEMK4UBxLzNyQU2z
< Accept-Ranges: bytes
< Content-Disposition: attachement; filename=rustup-init.sh
```

The `get_url` module fails with the following output:

```
Failed to validate the SSL certificate for sh.rustup.rs:443. Make sure your managed systems have a valid CA certificate installed. If the website serving the url uses SNI you need python >= 2.7.9 on your managed machine or you can install the `urllib3`, `pyopenssl`, `ndg-httpsclient`, and `pyasn1` python modules to perform SNI verification in python >= 2.6. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/pki/ca-trust/extracted/pem, /etc/pki/tls/certs, /usr/share/ca-certificates/cacert.org, /etc/ansible. The exception msg was: hostname 'sh.rustup.rs' doesn't match either of '*.rust-lang.org', 'rust-lang.org'.
```

All of the aforementioned packages have been installed via `pip`.

##### STEPS TO REPRODUCE

See [entire pull request](https://github.com/naftulikay/ansible-role-vagrant-rust-dev/pull/1). I have tested this problem with multiple Pythons:

 - 2.6
 - 2.7
 - 3.4
 - 3.5
 - 3.6

And on both the precise default Travis CI, and the beta trusty image.

##### EXPECTED RESULTS

I would expect it to work.

##### ACTUAL RESULTS

It doesn't work. Please see [the pull request](https://github.com/naftulikay/ansible-role-vagrant-rust-dev/pull/1) for an assortment of other information and actual Travis output for multiple Python versions, Travis CI images, etc.

**Comment by snowsky** [comment]
Run into a similar issue when using get_url with SSL. The workaround is to use shell module with curl or wget.

**Comment by snowsky** [comment]
Just tried and found that if validate_certs=False was added it can work.

**Comment by naftulikay** [comment]
Understood; I know I could use that workaround, but I'd like to stay within Ansible if possible and it doesn't look like I'm doing anything too out of the ordinary here.

**Comment by alikins** [comment]
Do other https urls fail to validate or only https://sh.rustup.rs?  Do other urls with 'Lets Encrypt' issued certificates fail? Do other urls using other CAs fail?

**Comment by brianshumate** [comment]
I've been seeing this same error for some time now, and IIRC its timing correlated with when `releases.hashicorp.com` began to require TLS v1.2+ for connections back in January so I cannot help but feel it's related to sites which are strict about requiring TLS v1.2 and higher or something similar.

I have seen past (possibly even current) Ansible issues where it incorrectly returns certificate related error messages for conditions (bad handshakes, etc.) which are not actually certificate issues so that could be muddying this up as well.

```
"Failed to validate the SSL certificate for releases.hashicorp.com:443. Make sure your managed systems have a valid CA certificate installed. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/ansible, /usr/local/etc/openssl"
```

The above error comes from a [`local_action` task calling `get_url`](https://github.com/brianshumate/ansible-consul/blob/master/tasks/install.yml#L39-L44) from macOS 10.12.4, Ansible 2.3.0.0, and Python 2.7.13 from PyEnv built with OpenSSL 1.0.2k from Homebrew, though this problem has been affecting me (and others) across multiple Ansible, Python, OpenSSL versions on macOS for many months now.

I can confirm that `releases.hashicorp.com` is otherwise fine and able to be accessed with other tools, e.g. `curl`, etc. A simple test with `openssl s_client -connect releases.hashicorp.com:443 -tls1_2` works fine from my Ansible control host as well.

Hopefully this provides some other data points, and let me know if there's anything specific that you'd like to see.

**Comment by sivel** [comment]
I am able to re-produce on my Mac using Python 2.7.10.  It seems as though Mac only has TLSv1.  Trying with a python3.5.2 install I have, which has TLSv1_2, it succeeds.

**Comment by brianshumate** [comment]
Although I did compile in an OpenSSL that should have 1.2 for my Python, I am going to switch to trying 3.5.2 as well.

**Comment by naftulikay** [comment]
Again, to reiterate, I'm running all this in Travis CI. With Python 2.7 and IIRC 3.4 locally it works just fine. I have tested on Travis' precise and trusty images with lots of versions of Python. I'm not sure how to debug further, is there a way that we can better provide debugging information relative to the environment to maybe catch a stacktrace and the real underlying failure at the TLS layer for validation/verification to succeed?

I think the answer has dawned on me: I'm thinking that the `get_url` module isn't setting server-name-indication and getting an alternate NGINX/Apache site and certificate. It appears that in my case, `sh.rustup.rs` is on the same server as something serving the following sites:

```
 hostname 'sh.rustup.rs' doesn't match either of '*.rust-lang.org', 'rust-lang.org'
```

@sivel does this make sense, could this be the problem?

**Comment by sivel** [comment]
That does not sound correct, at least not in my testing. I suppose you and I could be experiencing different issues, however in my tests it fails far before performing any SNI.  It fails to even establish an connection, seemingly because the TLS requirements exceed the capabilities of my python installation.

When using a python install with appropriate TLS capabilities, everything seemed to work as expected.  I have not tested with a similarly built py27, but the conclusion leads me to believe this may not be a situation that can be resolved within ansible.  It may be restricted to the capabilities of the installed python.

**Comment by naftulikay** [comment]
[Here][build] is a log of the Ansible output for my failure:

```
TASK [ansible-role-vagrant-rust-dev : download rustup] *************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to validate the SSL certificate for sh.rustup.rs:443. Make sure your managed systems have a valid CA certificate installed. If the website serving the url uses SNI you need python >= 2.7.9 on your managed machine or you can install the `urllib3`, `pyopenssl`, `ndg-httpsclient`, and `pyasn1` python modules to perform SNI verification in python >= 2.6. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/pki/ca-trust/extracted/pem, /etc/pki/tls/certs, /usr/share/ca-certificates/cacert.org, /etc/ansible"}
```

The linked build was on Python 2.7, but here are all of them:

 - 2.6: https://travis-ci.org/naftulikay/ansible-role-vagrant-rust-dev/jobs/222495981
 - 2.7: https://travis-ci.org/naftulikay/ansible-role-vagrant-rust-dev/jobs/222495982
 - 3.4: https://travis-ci.org/naftulikay/ansible-role-vagrant-rust-dev/jobs/222495983
 - 3.5: https://travis-ci.org/naftulikay/ansible-role-vagrant-rust-dev/jobs/222495984
 - 3.6: https://travis-ci.org/naftulikay/ansible-role-vagrant-rust-dev/jobs/222495985

If you check out [my linked pull request above][pr], you'll see that in both precise and trusty images in Travis, across all versions of Python, failure happens the same.

 [build]: https://travis-ci.org/naftulikay/ansible-role-vagrant-rust-dev/jobs/222495982
 [pr]: https://github.com/naftulikay/ansible-role-vagrant-rust-dev/pull/1

**Comment by naftulikay** [comment]
@sivel The reason I'm suspecting SNI is because I'm requesting `sh.rustup.rs` and it fails for me saying a mismatched certificate was returned and `sh.rustup.rs` is not covered by the certificate, which covers `*.rust-lang.org` and `rust-lang.org`.

### No SNI

If I disable SNI, then it becomes very apparent that certificate is not valid for `sh.rustup.rs`.

```
$ echo quit | openssl s_client -connect sh.rustup.rs:443 | head
depth=3 C = SE, O = AddTrust AB, OU = AddTrust External TTP Network, CN = AddTrust External CA Root
verify return:1
depth=2 C = GB, ST = Greater Manchester, L = Salford, O = COMODO CA Limited, CN = COMODO RSA Certification Authority
verify return:1
depth=1 C = GB, ST = Greater Manchester, L = Salford, O = COMODO CA Limited, CN = COMODO RSA Domain Validation Secure Server CA
verify return:1
depth=0 OU = Domain Control Validated, OU = EssentialSSL Wildcard, CN = *.rust-lang.org
verify return:1
CONNECTED(00000003)
---
Certificate chain
 0 s:/OU=Domain Control Validated/OU=EssentialSSL Wildcard/CN=*.rust-lang.org
   i:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Domain Validation Secure Server CA
 1 s:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Domain Validation Secure Server CA
   i:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Certification Authority
 2 s:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Certification Authority
   i:/C=SE/O=AddTrust AB/OU=AddTrust External TTP Network/CN=AddTrust External CA Root
---
```

### With SNI

With SNI enabled, I get the right certificate.

```
$ echo quit | openssl s_client -servername sh.rustup.rs -connect sh.rustup.rs:443 | head
depth=2 O = Digital Signature Trust Co., CN = DST Root CA X3
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
verify return:1
depth=0 CN = sh.rustup.rs
verify return:1
CONNECTED(00000003)
---
Certificate chain
 0 s:/CN=sh.rustup.rs
   i:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
DONE
 1 s:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
   i:/O=Digital Signature Trust Co./CN=DST Root CA X3
---
Server certificate
-----BEGIN CERTIFICATE-----
```

---

In summary, I still think this an SNI thing based on the errors I'm seeing.

**Comment by sivel** [comment]
Ansible supports SNI, assuming that either the python version supports it, or if you follow the instructions and install urllib3 and the other deps.

You can try to determine if the python version has the appropriate functionality, by using something like:

```
- debug: var=ansible_python
```

On my python version I see:

```
localhost | SUCCESS => {
    "ansible_facts": {
        "ansible_python": {
            "executable": "/usr/bin/python",
            "has_sslcontext": true,
            "type": "CPython",
            "version": {
                "major": 2,
                "micro": 10,
                "minor": 7,
                "releaselevel": "final",
                "serial": 0
            },
            "version_info": [
                2,
                7,
                10,
                "final",
                0
            ]
        }
    },
    "changed": false
}
```

The `"has_sslcontext": true,` indicates that it can utilize SNI.

However, as the error states, you could also install  `urllib3`, `pyopenssl`, `ndg-httpsclient`, and `pyasn1`, to get SNI functionality on python versions that do not natively support SNI.

**Comment by naftulikay** [comment]
@sivel I am installing those packages as well: https://github.com/naftulikay/ansible-role-vagrant-rust-dev/blob/a71b8e87d39444a62b53a7b1fc5d31dc82bb0dc9/requirements.txt

I will attempt to do what you suggested for debug info in Travis.

**Comment by mcapuccini** [comment]
I am experiencing the same issue with Ansible 2.2.0.0 and Ansible 2.2.1.0 when running from Travis. If I run the same on my laptop it works. I tried to install `ca-certificates`, and even to import the certificate manually, but no luck when running `get_url` from Travis.

```yaml
get_url:
        url: "https://httpd.pharmb.io/kubenow-img/{{ image_name }}.qcow2"
        dest: /tmp/{{ image_name }}.qcow2
```

```bash
TASK [download KubeNow image] **************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to validate the SSL certificate for httpd.pharmb.io:443. Make sure your managed systems have a valid CA certificate installed. If the website serving the url uses SNI you need python >= 2.7.9 on your managed machine or you can install the `urllib3`, `pyopenssl`, `ndg-httpsclient`, and `pyasn1` python modules to perform SNI verification in python >= 2.6. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/pki/ca-trust/extracted/pem, /etc/pki/tls/certs, /usr/share/ca-certificates/cacert.org, /etc/ansible"}
```

**Comment by sivel** [comment]
I'll have to see if I can do some digging on Travis to figure out what it's issue is.

**Comment by sivel** [comment]
I've got some good news, in getting `get_url` and `uri` to work on Travis CI while being able to validate certs.

There are a few things that you can do here:

1. Currently you have `localhost` explicitly defined in your inventory, which prevents ansible from using the implicit `localhost` which would set the correct python interpreter to utilize.  I'd recommend removing `localhost` from your inventory, and that would immediately solve your problem. As it stands now, the way you have it, causes ansible to use `/usr/bin/python` instead of the python from the travis virtual environment
1. You could define `ansible_python_interpreter` via the CLI for the ansible execution, so that it uses the correct python from the virtual environment in travis, instead of `/usr/bin/python`

(2) is my recommendation.

It looks like by default, the global version of python installed in the travis environment is 2.7.3, which does not have SSLContext and does not natively support SNI.

However, the python versions in the virtualenvs that you should be using, do actually have the necessary SSL functionality to handle this.

Using the following `.travis.yml` contents, while setting the appropriate `ansible_python_interpreter` resolves the issue for me:

```
language: python

matrix:
  include:
    - python: 2.7
      env: PYBIN=/home/travis/virtualenv/python2.7.12/bin/python2.7
    - python: 3.5
      env: PYBIN=/home/travis/virtualenv/python3.5.2/bin/python3.5
    - python: 3.6
      env: PYBIN=/home/travis/virtualenv/python3.6.1/bin/python3.6

script:
  - pip install ansible
  - ansible all -i localhost, -c local -m setup -a 'filter=ansible_python' -e ansible_python_interpreter=$PYBIN
  - ansible all -i localhost, -c local -m uri -a url=https://sh.rustup.rs/ -e ansible_python_interpreter=$PYBIN
```

Results: https://travis-ci.org/sivel/travis-test/builds/226470804

An additional test without explicitly defining localhost in inventory to illustrate (2):

https://github.com/sivel/travis-test/tree/456249c49cd9f69bf9056ea2d7571d69e1049f1b
https://travis-ci.org/sivel/travis-test/builds/226480657

Using implicit `localhost`, will require you to remove your inventory file, not specify an inventory file, and explicitly use `localhost` for your `hosts` in the playbook.

Both options work in py2.7, py3.5 and py3.6.  Ansible does not support python3.4, and I just didn't test with py2.6, but using the correct py2.6, which should be 2.6.9 on Travis, should also succeed.

I'm going to go ahead and close this out.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by naftulikay** [comment]
@sivel thank you so much!
